src/conn/conn_compact.c

```diff
@@ -0,0 +1,260 @@
+
+    if (0) {
+err:
+        WT_IGNORE_RET(__wt_panic(session, ret, "compact server error"));
+    }
+    return (WT_THREAD_RET_VALUE);
+}
+
+/*
+ * __wt_compact_server_create --
+ *     Start the compact thread.
+ */
+int
+__wt_compact_server_create(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    uint32_t session_flags;
+
+    conn = S2C(session);
+
+    /* Set first, the thread might run before we finish up. */
+    FLD_SET(conn->server_flags, WT_CONN_SERVER_COMPACT);
+
+    /*
+     * Compaction does enough I/O it may be called upon to perform slow operations for the block
+     * manager.
+     */
+    session_flags = WT_SESSION_CAN_WAIT;
+    WT_RET(__wt_open_internal_session(
+      conn, "compact-server", true, session_flags, 0, &conn->background_compact.session));
+    session = conn->background_compact.session;
+
+    WT_RET(__wt_cond_alloc(session, "compact server", &conn->background_compact.cond));
+
+    WT_RET(__wt_thread_create(session, &conn->background_compact.tid, __compact_server, session));
+    conn->background_compact.tid_set = true;
+
+    return (0);
+}
+
+/*
+ * __wt_compact_server_destroy --
+ *     Destroy the background compaction server thread.
+ */
+int
+__wt_compact_server_destroy(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+
+    conn = S2C(session);
+
+    FLD_CLR(conn->server_flags, WT_CONN_SERVER_COMPACT);
+    if (conn->background_compact.tid_set) {
+        conn->background_compact.running = false;
+        __wt_cond_signal(session, conn->background_compact.cond);
+        WT_TRET(__wt_thread_join(session, &conn->background_compact.tid));
+        conn->background_compact.tid_set = false;
+    }
+    __wt_cond_destroy(session, &conn->background_compact.cond);
+
+    /* Close the server thread's session. */
+    if (conn->background_compact.session != NULL) {
+        WT_TRET(__wt_session_close_internal(conn->background_compact.session));
+        conn->background_compact.session = NULL;
+    }
+
+    return (ret);
+}
+
+/*
+ * __wt_compact_signal --
+ *     Signal the compact thread. Return EBUSY if the background compaction server has not processed
+ *     a previous signal yet.
+ */
+int
+__wt_compact_signal(WT_SESSION_IMPL *session, const char *config)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    bool running;
+
+    conn = S2C(session);
+
+    /* Wait for any previous signal to be processed first. */
+    if (conn->background_compact.signalled)
+        return (EBUSY);
+
+    /*
+     * FIXME-WT-11334: Using the config argument, check whether the background compaction server is
+     * already enabled/disabled and update the parameters accordingly.
+     */
+
+    __wt_spin_lock(session, &conn->background_compact.lock);
+    running = conn->background_compact.running;
+    conn->background_compact.running = !running;
+    __wt_free(session, conn->background_compact.config);
+    WT_ERR(__wt_strndup(
```

Comment:
We need to do a bit more in terms of copying the configuration here. We don't want to copy the entire config string eg. "background=true,free_space_target=20MB". We need to strip out the background component and only copy the rest of the configurations to avoid session::compact ending up back here on an infinite loop.

Reply:
From the background compact, we pass the `uri` to the compact along with the copied config. As the config has the background option, it will fail. This needs to be addressed.

Reply:
@kommiharibabu, this will be checked by WT-11334 (#9461), see [here](https://github.com/wiredtiger/wiredtiger/pull/9461/files#diff-eab825f891ddcec54bad4fdc922828201330ef985f8458558ee6e20615658a75R338-R343).

When this function is executed, the expected configuration is `"background=on/off,free_space_target=xx,timeout=yy"`. Note that the only mandatory item is `background=on/off`. See my latest changes 049eb0ccee1258ad0f5cba2c794bc09dc8724517, especially [this line](https://github.com/wiredtiger/wiredtiger/pull/9456/commits/049eb0ccee1258ad0f5cba2c794bc09dc8724517#diff-ad85b77b14594b779f82d482423c144c3287a45c607d76e7a68db8466ac93e30R244).

Reply:
Additional changes can be seen in ddc9e914caa4239ceca35ab7b5021da742be7205, the initial solution was not working correctly.